lessons/misc/quantum-computing/Grover's Algorthim with 2 Qubits.ipynb

```diff
@@ -0,0 +1,396 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Simulating Grover's Search Algorithm with 2 Qubits"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 2,
+   "metadata": {
+    "collapsed": true
+   },
+   "outputs": [],
+   "source": [
+    "import numpy as np\n",
+    "from matplotlib import pyplot as plt\n",
+    "%matplotlib inline"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Define the zero and one vectors\n",
+    "Define the initial state $\\psi$"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 22,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "[[1]\n",
+      " [0]\n",
+      " [0]\n",
+      " [0]]\n"
+     ]
+    }
+   ],
+   "source": [
+    "zero = np.matrix([[1],[0]]);\n",
+    "one = np.matrix([[0],[1]]);\n",
+    "\n",
+    "psi = np.kron(zero,zero);\n",
+    "print(psi)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Define the gates we will use:\n",
+    "\n",
+    "$\n",
+    "\\text{Id} = \\begin{pmatrix} \n",
+    "1 & 0  \\\\\n",
+    "0 & 1  \n",
+    "\\end{pmatrix},\n",
+    "\\quad\n",
+    "X = \\begin{pmatrix} \n",
+    "0 & 1   \\\\\n",
+    "1 & 0  \n",
+    "\\end{pmatrix},\n",
+    "\\quad\n",
+    "Z = \\begin{pmatrix} \n",
+    "1 &  0 \\\\\n",
+    "0 & -1 \n",
+    "\\end{pmatrix},\n",
+    "\\quad\n",
+    "H = \\frac{1}{\\sqrt{2}}\\begin{pmatrix} \n",
+    "1 & 1   \\\\\n",
+    "1 & -1  \n",
+    "\\end{pmatrix},\n",
+    "\\quad\n",
+    "\\text{CNOT} = \\begin{pmatrix} \n",
+    "1 & 0 & 0 & 0 \\\\\n",
+    "0 & 1 & 0 & 0 \\\\  \n",
+    "0 & 0 & 0 & 1 \\\\\n",
+    "0 & 0 & 1 & 0\n",
+    "\\end{pmatrix},\n",
+    "\\quad\n",
+    "CZ = (\\text{Id} \\otimes H) \\text{ CNOT } (\\text{Id} \\otimes H) \n",
+    "$"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 23,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "[[ 1.  0.  0.  0.]\n",
+      " [ 0.  1.  0.  0.]\n",
+      " [ 0.  0.  1.  0.]\n",
+      " [ 0.  0.  0. -1.]]\n"
+     ]
+    }
+   ],
+   "source": [
+    "Id = np.matrix([[1,0],[0,1]]);\n",
+    "X = np.matrix([[0,1],[1,0]]);\n",
+    "Z = np.matrix([[1,0],[0,-1]]);\n",
+    "H = np.sqrt(0.5) * np.matrix([[1,1],[1,-1]]);\n",
+    "\n",
+    "CNOT = np.matrix([[1,0,0,0],[0,1,0,0],[0,0,0,1],[0,0,1,0]]);\n",
+    "CZ = np.kron(Id,H).dot(CNOT).dot(np.kron(Id,H));\n",
```

Comment:
On Python 3.5+, you can do matrix multiplication with `@`: `CZ = np.kron(Id, H) @ CNOT @ np.kron(ID, H)`